File: functions/team_access.js
'use strict';

// Keep this policy aligned with lib/models/team_access.dart; parity is tested.
const permissions = Object.freeze([
  'manageOrganization', 'manageTeam', 'manageProperty', 'manageLease',
  'readBookings', 'manageBookings', 'collectPayments', 'overridePrices',
  'refundPayments', 'readFinancialReports', 'readOwnActivity', 'readAllActivity',
  'exportData', 'importData', 'connectDrive', 'readAssignedTasks', 'updateAssignedTasks',
]);
const roles = Object.freeze({
  owner: permissions,
  administrator: permissions,
  manager: ['manageProperty', 'manageLease', 'readBookings', 'manageBookings',
    'collectPayments', 'overridePrices', 'readFinancialReports', 'readOwnActivity'],
  receptionist: ['readBookings', 'manageBookings', 'collectPayments', 'readOwnActivity'],
  housekeeper: ['readAssignedTasks', 'updateAssignedTasks', 'readOwnActivity'],
  accountant: ['readBookings', 'collectPayments', 'refundPayments', 'readFinancialReports', 'readOwnActivity'],
  viewer: ['readFinancialReports'],
});
const overrides = Object.freeze(['overridePrices', 'refundPayments', 'exportData', 'importData']);
const knownRole = role => Object.hasOwn(roles, role);

function allows(membership, permission, {organizationId, userId, buildingId} = {}) {
  if (!membership || membership.accessVersion !== 2 || membership.status !== 'active' ||
      !['all', 'selected'].includes(membership.buildingScope) ||
      !knownRole(membership.role) || !permissions.includes(permission)) return false;
  if (organizationId !== undefined && membership.organizationId !== organizationId) return false;
  if (userId !== undefined && membership.ownerId !== userId) return false;
  if (buildingId !== undefined && (typeof buildingId !== 'string' || !buildingId ||
      (membership.buildingScope !== 'all' &&
        !(membership.buildingScope === 'selected' && Array.isArray(membership.buildingIds) && membership.buildingIds.includes(buildingId))))) return false;
  const override = membership.permissionOverrides?.[permission];
  return overrides.includes(permission) && typeof override === 'boolean'
    ? override : roles[membership.role].includes(permission);
}

function canManageAccessOf(actor, targetRole, context) {
  return allows(actor, 'manageTeam', context) && targetRole !== 'owner' &&
    (actor.role === 'owner' || targetRole !== 'administrator');
}

// Produces a proposal only. It never writes or silently assigns legacy members.
function migrationProposal(organization, memberships) {
  const issues = [];
  const seen = new Set();
  const proposals = memberships.map(m => {
    if (m.organizationId !== organization.id) {
      issues.push({membershipId: m.id, reason: 'organizationMismatch'});
      return {membershipId: m.id, action: 'review'};
    }
    if (!m.ownerId || seen.has(m.ownerId)) {
      issues.push({membershipId: m.id, reason: 'missingOrDuplicateAccount'});
      return {membershipId: m.id, action: 'review'};
    }
    seen.add(m.ownerId);
    if (m.accessVersion === 2) return {membershipId: m.id, action: 'unchanged'};
    const owner = m.ownerId === organization.createdBy;
    const role = owner ? 'owner' : m.role === 'admin' ? 'administrator' : null;
    const status = m.status === 'active' ? (role ? 'active' : 'assignmentRequired') : 'suspended';
    return {membershipId: m.id, action: 'review', proposed: {
      accessVersion: 2, role, status, buildingScope: role ? 'all' : 'selected',
      buildingIds: [], permissionOverrides: {},
    }};
  });
  const creator = memberships.find(m => m.organizationId === organization.id && m.ownerId === organization.createdBy);
  if (!creator || creator.status !== 'active') issues.push({reason: 'missingActiveOwner'});
  return {organizationId: organization.id, proposals, issues};
}

module.exports = {permissions, roles, overrides, allows, canManageAccessOf, migrationProposal};

File: lib/models/team_access.dart
/// Versioned access policy. Legacy roles deliberately grant no v2 access.
enum TeamRole { owner, administrator, manager, receptionist, housekeeper, accountant, viewer }

enum TeamPermission {
  manageOrganization, manageTeam, manageProperty, manageLease,
  readBookings, manageBookings, collectPayments, overridePrices, refundPayments,
  readFinancialReports, readOwnActivity, readAllActivity, exportData, importData,
  connectDrive, readAssignedTasks, updateAssignedTasks,
}

class TeamAccess {
  static const version = 2;
  static const overridable = {
    TeamPermission.overridePrices, TeamPermission.refundPayments,
    TeamPermission.exportData, TeamPermission.importData,
  };
  final TeamRole? role;
  final String status;
  final bool allBuildings;
  final Set<String> buildingIds;
  final Map<TeamPermission, bool> overrides;

  TeamAccess({required this.role, required this.status,
    this.allBuildings = false, Set<String> buildingIds = const {},
    Map<TeamPermission, bool> overrides = const {}})
      : buildingIds = Set.unmodifiable(buildingIds),
        overrides = Map.unmodifiable(overrides);

  static Set<TeamPermission> defaults(TeamRole? role) {
    switch (role) {
      case TeamRole.owner:
      case TeamRole.administrator:
        return TeamPermission.values.toSet();
      case TeamRole.manager:
        return {TeamPermission.manageProperty, TeamPermission.manageLease,
          TeamPermission.readBookings, TeamPermission.manageBookings,
          TeamPermission.collectPayments, TeamPermission.overridePrices,
          TeamPermission.readFinancialReports, TeamPermission.readOwnActivity};
      case TeamRole.receptionist:
        return {TeamPermission.readBookings, TeamPermission.manageBookings,
          TeamPermission.collectPayments, TeamPermission.readOwnActivity};
      case TeamRole.housekeeper:
        return {TeamPermission.readAssignedTasks, TeamPermission.updateAssignedTasks,
          TeamPermission.readOwnActivity};
      case TeamRole.accountant:
        return {TeamPermission.readBookings, TeamPermission.collectPayments,
          TeamPermission.refundPayments, TeamPermission.readFinancialReports,
          TeamPermission.readOwnActivity};
      case TeamRole.viewer:
        return {TeamPermission.readFinancialReports};
      case null:
        return {};
    }
  }

  bool allows(TeamPermission permission, {String? buildingId}) {
    if (status != 'active' || role == null) return false;
    if (buildingId != null &&
        (buildingId.isEmpty || (!allBuildings && !buildingIds.contains(buildingId)))) {
      return false;
    }
    return overridable.contains(permission) && overrides.containsKey(permission)
        ? overrides[permission]!
        : defaults(role).contains(permission);
  }

  /// Call separately from manageTeam: administrators cannot alter peers/owners.
  bool canManageAccessOf(TeamRole? target) =>
      allows(TeamPermission.manageTeam) && target != TeamRole.owner &&
      (role == TeamRole.owner || target != TeamRole.administrator);

  factory TeamAccess.fromMap(Map<String, dynamic> map) {
    final validVersion = map['accessVersion'] == version &&
        (map['buildingScope'] == 'all' || map['buildingScope'] == 'selected');
    final role = TeamRole.values.where((r) => r.name == map['role']).firstOrNull;
    final rawOverrides = map['permissionOverrides'];
    return TeamAccess(
      role: validVersion ? role : null,
      status: map['status'] is String ? map['status'] as String : 'assignmentRequired',
      allBuildings: validVersion && map['buildingScope'] == 'all',
      buildingIds: map['buildingIds'] is List
          ? (map['buildingIds'] as List).whereType<String>().where((s) => s.isNotEmpty).toSet() : {},
      overrides: rawOverrides is Map ? {
        for (final p in overridable)
          if (rawOverrides[p.name] is bool) p: rawOverrides[p.name] as bool,
      } : {},
    );
  }

  Map<String, dynamic> toMap() => {
    'accessVersion': version, 'role': role?.name,
    'status': status, 'buildingScope': allBuildings ? 'all' : 'selected',
    'buildingIds': buildingIds.toList()..sort(),
    'permissionOverrides': {for (final e in overrides.entries)
      if (overridable.contains(e.key)) e.key.name: e.value},
  };
}
